File: agent/runner.py
"""Agent main loop: TOFU register → wait approval → heartbeat + fetch-certs.

Lifecycle:
  1. Generate RSA key pair (if not exists), compute fingerprint
  2. If not registered (no agent_token on disk):
     a. POST /register with {name, fingerprint}
     b. Poll /register/status until approved, save agent_token to disk
  3. Main loop (heartbeat_interval):
     a. POST /heartbeat (keeps last_seen fresh)
     b. For each entry in cert_table:
        - Read local cert file (if exists), extract not_after
        - Build batch check payload
     c. POST /fetch-certs with batch
     d. Deploy any updates (write cert/key files, reload nginx)
"""

import logging
import time
from datetime import datetime, timezone
from pathlib import Path

import httpx
from cryptography import x509

from agent.client import ControlPlaneClient
from agent.config import AgentConfig
from agent.crypto import compute_fingerprint, generate_key_pair, load_private_key
from agent.deployer import deploy_to_nginx

logger = logging.getLogger(__name__)


def run(config: AgentConfig) -> None:
    """Entry point – blocks forever."""
    config.state_dir.mkdir(parents=True, exist_ok=True)

    # ------------------------------------------------------------------
    # Step 1: Ensure we have a key pair
    # ------------------------------------------------------------------
    if config.key_path.exists():
        key = load_private_key(config.key_path)
        logger.info("Loaded existing private key from %s", config.key_path)
    else:
        key = generate_key_pair(config.key_path)
        logger.info("Generated new private key at %s", config.key_path)

    fingerprint = compute_fingerprint(key)
    logger.info("Agent fingerprint: %s", fingerprint)

    client = ControlPlaneClient(config)

    # ------------------------------------------------------------------
    # Step 2: Register (TOFU) if no agent_token
    # ------------------------------------------------------------------
    if not config.is_registered():
        _do_register(config, client, fingerprint)
    else:
        logger.info("Already registered, agent_token loaded from disk")

    # ------------------------------------------------------------------
    # Step 3: Main heartbeat + fetch-certs loop
    # ------------------------------------------------------------------
    logger.info(
        "Entering main loop (interval=%ds), %d cert(s) in table",
        config.heartbeat_interval,
        len(config.cert_table),
    )

    while True:
        try:
            # Heartbeat
            client.heartbeat()
            logger.debug("Heartbeat OK")

            # Fetch and deploy certs
            _do_fetch_and_deploy(config, client)

        except httpx.HTTPStatusError as e:
            if e.response.status_code in (401, 403):
                logger.warning(
                    "Agent token rejected (HTTP %d) – clearing token and re-registering",
                    e.response.status_code,
                )
                config.agent_token = ""
                if config.agent_token_path.exists():
                    config.agent_token_path.unlink()
                _do_register(config, client, fingerprint)
            else:
                logger.exception("HTTP error in main loop (status %d)", e.response.status_code)
        except Exception:
            logger.exception("Error in main loop, will retry")

        time.sleep(config.heartbeat_interval)


# ---------------------------------------------------------------------------
# Registration (TOFU)
# ---------------------------------------------------------------------------


def _do_register(
    config: AgentConfig,
    client: ControlPlaneClient,
    fingerprint: str,
) -> None:
    """Submit TOFU registration and poll until approved."""
    logger.info("Starting TOFU registration for '%s'", config.agent_name)

    # Submit registration
    resp = client.register(fingerprint)
    reg_status = resp.get("status")
    agent_id = resp.get("agent_id")

    logger.info("Registration response: status=%s, agent_id=%s", reg_status, agent_id)

    if reg_status == "approved" and resp.get("agent_token"):
        # Already approved (re-registration case)
        config.save_agent_token(resp["agent_token"])
        logger.info("Agent token received immediately")
        return

    if reg_status == "pending":
        # Poll until approved
        logger.info(
            "Registration pending admin approval. "
            "Polling every %ds...",
            config.poll_interval,
        )
        _poll_until_approved(config, client, agent_id, fingerprint)
        return

    raise RuntimeError(f"Unexpected registration status: {reg_status}")


def _poll_until_approved(
    config: AgentConfig,
    client: ControlPlaneClient,
    agent_id: str,
    fingerprint: str,
) -> None:
    """Block until admin approves this agent, saving agent_token when done."""
    while True:
        try:
            resp = client.check_registration_status(agent_id, fingerprint)
            poll_status = resp.get("status")

            if poll_status == "approved" and resp.get("agent_token"):
                config.save_agent_token(resp["agent_token"])
                logger.info("Admin approved! Agent token saved.")
                return

            if poll_status == "rejected":
                raise RuntimeError("Agent registration was rejected by admin")

            logger.debug("Still waiting for admin approval (status=%s)", poll_status)

        except httpx.HTTPStatusError as e:
            logger.warning("Polling failed (HTTP %d), will retry", e.response.status_code)
        except Exception:
            logger.exception("Unexpected error while polling approval, will retry")

        time.sleep(config.poll_interval)


# ---------------------------------------------------------------------------
# Fetch and deploy certificates
# ---------------------------------------------------------------------------


def _do_fetch_and_deploy(config: AgentConfig, client: ControlPlaneClient) -> None:
    """Read local cert table, ask platform for updates, deploy any changes."""
    if not config.cert_table:
        logger.debug("cert_table is empty, nothing to check")
        return

    # Build batch check payload
    cert_checks = []
    for entry in config.cert_table:
        local_path = entry.get("local_path", "")
        if not local_path:
            continue

        current_not_after = _read_cert_not_after(local_path)
        cert_checks.append({
            "local_path": local_path,
            "current_not_after": current_not_after,
        })

    if not cert_checks:
        return

    # Ask the control plane
    try:
        resp = client.fetch_certs(cert_checks)
    except Exception:
        logger.exception("Failed to fetch certs from control plane")
        return

    # Deploy updates
    for update in resp.get("updates", []):
        if not update.get("has_update"):
            continue

        local_path = update["local_path"]
        logger.info("Cert update available for %s", local_path)
        _deploy_cert_update(config, update)


def _read_cert_not_after(local_path: str) -> str | None:
    """Read a PEM cert file and return its not_after as ISO 8601 string, or None."""
    try:
        p = Path(local_path)
        if not p.exists():
            return None
        cert_pem = p.read_bytes()
        cert = x509.load_pem_x509_certificate(cert_pem)
        return cert.not_valid_after_utc.isoformat()
    except Exception:
        logger.warning("Failed to read/parse cert at %s", local_path)
        return None


def _deploy_cert_update(config: AgentConfig, update: dict) -> None:
    """Write cert+key files to local_path and reload nginx."""
    local_path = update["local_path"]
    cert_pem = update.get("cert_pem")
    key_pem = update.get("key_pem")
    chain_pem = update.get("chain_pem")

    p = Path(local_path)
    # Derive key and chain paths from cert path convention
    # e.g. /etc/nginx/ssl/api.crt → /etc/nginx/ssl/api.key
    key_path = p.with_suffix(".key")
    chain_path = p.with_suffix(".chain.crt")

    # Backup existing files
    old_cert = p.with_suffix(".crt.old")
    old_key = key_path.with_suffix(".key.old")

    if p.exists():
        _backup(p, old_cert)
    if key_path.exists():
        _backup(key_path, old_key)

    try:
        p.parent.mkdir(parents=True, exist_ok=True)

        if cert_pem:
            p.write_text(cert_pem)
            p.chmod(0o644)

        if key_pem:
            key_path.write_text(key_pem)
            key_path.chmod(0o600)

        if chain_pem:
            chain_path.write_text(chain_pem)
            chain_path.chmod(0o644)

        # Reload nginx
        import subprocess
        if config.nginx_reload_cmd:
            result = subprocess.run(
                config.nginx_reload_cmd,
                shell=True,
                capture_output=True,
                text=True,
            )
            if result.returncode != 0:
                raise RuntimeError(f"nginx reload failed: {result.stderr}")

        logger.info("Deployed cert update to %s", local_path)

        # Clean up backups
        old_cert.unlink(missing_ok=True)
        old_key.unlink(missing_ok=True)

    except Exception:
        logger.exception("Failed to deploy cert to %s, restoring backups", local_path)
        if old_cert.exists():
            _restore(old_cert, p)
        if old_key.exists():
            _restore(old_key, key_path)
        raise


def _backup(src: Path, dst: Path) -> None:
    dst.write_bytes(src.read_bytes())
    dst.chmod(src.stat().st_mode)


def _restore(backup: Path, target: Path) -> None:
    target.write_bytes(backup.read_bytes())
    target.chmod(backup.stat().st_mode)
    backup.unlink(missing_ok=True)

File: agent-go/internal/runner/runner.go
package runner

import (
	"crypto/x509"
	"encoding/pem"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"go.uber.org/zap"

	"github.com/cert-control-plane/agent-go/internal/client"
	"github.com/cert-control-plane/agent-go/internal/config"
	"github.com/cert-control-plane/agent-go/internal/crypto"
)

// Runner manages the agent lifecycle
type Runner struct {
	config *config.Config
	logger *zap.Logger
	client *client.ControlPlaneClient
}

// New creates a new Runner
func New(cfg *config.Config, logger *zap.Logger) *Runner {
	return &Runner{
		config: cfg,
		logger: logger,
	}
}

// Run starts the agent main loop
func (r *Runner) Run() error {
	r.logger.Info("cert-agent starting",
		zap.String("name", r.config.AgentName),
		zap.String("cp", r.config.ControlPlaneURL),
	)

	// Ensure state directory exists
	if err := os.MkdirAll(r.config.StateDir, 0755); err != nil {
		return fmt.Errorf("failed to create state dir: %w", err)
	}

	// Ensure we have a key pair
	fingerprint, err := r.ensureKeyPair()
	if err != nil {
		return fmt.Errorf("failed to ensure key pair: %w", err)
	}

	// Create HTTP client (no mTLS)
	c, err := client.New(r.config)
	if err != nil {
		return fmt.Errorf("failed to create client: %w", err)
	}
	r.client = c

	// Register if needed
	if !r.config.IsRegistered() {
		if err := r.register(fingerprint); err != nil {
			return fmt.Errorf("registration failed: %w", err)
		}
	} else {
		r.logger.Info("Already registered")
	}

	return r.runMainLoop()
}

// ensureKeyPair loads or generates the RSA key pair, returning the fingerprint
func (r *Runner) ensureKeyPair() (string, error) {
	keyPath := r.config.KeyPath()

	if _, err := os.Stat(keyPath); os.IsNotExist(err) {
		r.logger.Info("Generating new key pair", zap.String("path", keyPath))
		if _, err := crypto.GenerateKeyPair(3072); err != nil {
			return "", fmt.Errorf("failed to generate key: %w", err)
		}
		if err := crypto.SavePrivateKey(nil, keyPath); err != nil {
			return "", fmt.Errorf("failed to save key: %w", err)
		}
	}

	keyPEM, err := os.ReadFile(keyPath)
	if err != nil {
		return "", fmt.Errorf("failed to read key: %w", err)
	}

	fingerprint, err := client.ComputeFingerprint(keyPEM)
	if err != nil {
		return "", fmt.Errorf("failed to compute fingerprint: %w", err)
	}

	r.logger.Info("Agent fingerprint", zap.String("fingerprint", fingerprint))
	return fingerprint, nil
}

// register performs TOFU registration and waits for admin approval
func (r *Runner) register(fingerprint string) error {
	r.logger.Info("Starting TOFU registration", zap.String("name", r.config.AgentName))

	resp, err := r.client.Register(fingerprint)
	if err != nil {
		return fmt.Errorf("registration request failed: %w", err)
	}

	r.logger.Info("Registration response",
		zap.String("status", resp.Status),
		zap.String("agent_id", resp.AgentID),
	)

	if resp.Status == "approved" && resp.AgentToken != nil {
		return r.saveToken(*resp.AgentToken)
	}

	if resp.Status == "pending" {
		return r.pollUntilApproved(resp.AgentID, fingerprint)
	}

	return fmt.Errorf("unexpected registration status: %s", resp.Status)
}

func (r *Runner) pollUntilApproved(agentID, fingerprint string) error {
	r.logger.Info("Waiting for admin approval", zap.Duration("poll_interval", r.config.PollInterval))

	for {
		resp, err := r.client.CheckRegistrationStatus(agentID, fingerprint)
		if err != nil {
			r.logger.Warn("Approval poll failed", zap.Error(err))
		} else {
			switch resp.Status {
			case "approved":
				if resp.AgentToken != nil {
					r.logger.Info("Admin approved!")
					return r.saveToken(*resp.AgentToken)
				}
			case "rejected":
				return fmt.Errorf("agent registration rejected by admin")
			default:
				r.logger.Debug("Still waiting", zap.String("status", resp.Status))
			}
		}
		time.Sleep(r.config.PollInterval)
	}
}

func (r *Runner) saveToken(token string) error {
	tokenPath := r.config.AgentTokenPath()
	if err := os.WriteFile(tokenPath, []byte(token), 0600); err != nil {
		return fmt.Errorf("failed to save agent token: %w", err)
	}
	r.config.AgentToken = token
	r.logger.Info("Agent token saved", zap.String("path", tokenPath))
	return nil
}

func (r *Runner) runMainLoop() error {
	r.logger.Info("Entering main loop",
		zap.Duration("interval", r.config.HeartbeatInterval),
		zap.Int("cert_table_size", len(r.config.CertTable)),
	)

	ticker := time.NewTicker(r.config.HeartbeatInterval)
	defer ticker.Stop()

	for {
		// Heartbeat
		if _, err := r.client.Heartbeat(); err != nil {
			r.logger.Error("Heartbeat failed", zap.Error(err))
		} else {
			r.logger.Debug("Heartbeat OK")

			// Fetch and deploy certs
			if err := r.fetchAndDeployCerts(); err != nil {
				r.logger.Error("Cert fetch/deploy failed", zap.Error(err))
			}
		}

		<-ticker.C
	}
}

func (r *Runner) fetchAndDeployCerts() error {
	if len(r.config.CertTable) == 0 {
		r.logger.Debug("cert_table is empty, nothing to check")
		return nil
	}

	// Build batch check payload
	var checks []client.CertCheckItem
	for _, entry := range r.config.CertTable {
		check := client.CertCheckItem{LocalPath: entry.LocalPath}

		// Read current cert's not_after if it exists
		if notAfter, err := readCertNotAfter(entry.LocalPath); err == nil && notAfter != "" {
			check.CurrentNotAfter = &notAfter
		}

		checks = append(checks, check)
	}

	resp, err := r.client.FetchCerts(checks)
	if err != nil {
		return fmt.Errorf("fetch-certs request failed: %w", err)
	}

	for _, update := range resp.Updates {
		if !update.HasUpdate {
			continue
		}
		r.logger.Info("Cert update available", zap.String("path", update.LocalPath))
		if err := r.deployCertUpdate(update); err != nil {
			r.logger.Error("Failed to deploy cert", zap.String("path", update.LocalPath), zap.Error(err))
		}
	}

	return nil
}

func readCertNotAfter(certPath string) (string, error) {
	data, err := os.ReadFile(certPath)
	if err != nil {
		return "", err
	}
	block, _ := pem.Decode(data)
	if block == nil {
		return "", fmt.Errorf("failed to decode PEM")
	}
	cert, err := x509.ParseCertificate(block.Bytes)
	if err != nil {
		return "", err
	}
	return cert.NotAfter.UTC().Format(time.RFC3339), nil
}

func (r *Runner) deployCertUpdate(update client.CertUpdateItem) error {
	certPath := update.LocalPath
	keyPath := replaceExt(certPath, ".crt", ".key")
	chainPath := replaceExt(certPath, ".crt", ".chain.crt")

	// Backup existing files
	oldCertPath := certPath + ".old"
	oldKeyPath := keyPath + ".old"
	_ = copyFile(certPath, oldCertPath)
	_ = copyFile(keyPath, oldKeyPath)

	// Write new files
	if err := os.MkdirAll(filepath.Dir(certPath), 0755); err != nil {
		return err
	}

	if update.CertPEM != nil {
		if err := os.WriteFile(certPath, []byte(*update.CertPEM), 0644); err != nil {
			restoreBackup(oldCertPath, certPath)
			return fmt.Errorf("failed to write cert: %w", err)
		}
	}

	if update.KeyPEM != nil {
		if err := os.WriteFile(keyPath, []byte(*update.KeyPEM), 0600); err != nil {
			restoreBackup(oldCertPath, certPath)
			return fmt.Errorf("failed to write key: %w", err)
		}
	}

	if update.ChainPEM != nil {
		if err := os.WriteFile(chainPath, []byte(*update.ChainPEM), 0644); err != nil {
			r.logger.Warn("Failed to write chain", zap.Error(err))
		}
	}

	// Reload nginx
	if r.config.NginxReloadCmd != "" {
		cmd := exec.Command("sh", "-c", r.config.NginxReloadCmd)
		if output, err := cmd.CombinedOutput(); err != nil {
			restoreBackup(oldCertPath, certPath)
			restoreBackup(oldKeyPath, keyPath)
			return fmt.Errorf("nginx reload failed: %w - %s", err, string(output))
		}
	}

	// Clean up backups
	os.Remove(oldCertPath)
	os.Remove(oldKeyPath)

	r.logger.Info("Cert deployed", zap.String("path", certPath))
	return nil
}

func replaceExt(path, oldExt, newExt string) string {
	n := len(path)
	if n >= len(oldExt) && path[n-len(oldExt):] == oldExt {
		return path[:n-len(oldExt)] + newExt
	}
	return path + newExt
}

func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, data, 0644)
}

func restoreBackup(backup, target string) {
	if _, err := os.Stat(backup); err == nil {
		_ = copyFile(backup, target)
	}
}
